schedule/scheduler.go

```diff
@@ -14,159 +18,301 @@ import (
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
+	// used as context for jobs submitted, ensure they are canceled when context is canceled.
+	s.context = ctx
 
-	taskComplete := make(chan struct{}, len(s.tasks))
-	tasksRunning := len(s.tasks)
-
-	for _, tc := range s.tasks {
-		go func(tc TaskConfig) {
-			// Report task is complete when this goroutine exits
-			defer func() {
-				taskComplete <- struct{}{}
-			}()
-
-			// Attempt to get the task lock if specified
-			if tc.Locker != nil {
-				if err := tc.Locker.Lock(ctx); err != nil {
-					if errors.Is(err, storage.ErrLockNotAcquired) {
-						log.Infow("task not started: lock not acquired", "task", tc.Name)
-						return
-					}
-					log.Errorw("task not started: lock not acquired", "task", tc.Name, "error", err.Error())
-					return
-				}
-				defer func() {
-					if err := tc.Locker.Unlock(ctx); err != nil {
-						if !errors.Is(err, context.Canceled) {
-							log.Errorw("failed to unlock task", "task", tc.Name, "error", err.Error())
-						}
-					}
-				}()
+	for _, tc := range s.jobs {
+		go s.execute(tc, s.scheduledJobComplete)
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		// A little jitter between scheduledTasks to reduce thundering herd effects on api.
+		wait.SleepWithJitter(s.jobDelay, 2)
+	}
+
+	// Wait until the context is done and handle new jobs as they are submitted.
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case newTask := <-s.jobQueue:
+			newTask.log = log.With("id", newTask.id, "name", newTask.Name)
+			s.jobs[newTask.id] = newTask
+			newTask.log.Infow("new job received")
+			go s.execute(newTask, s.workerJobComplete)
+		case <-s.scheduledJobComplete:
+			// A job has completed
+			s.scheduledJobsRunning--
+			if s.scheduledJobsRunning == 0 {
+				log.Info("no scheduled jobs running")
+			}
+		case <-s.workerJobComplete:
+			s.workerJobsRunning--
+			if s.workerJobsRunning == 0 {
+				log.Info("no worker jobs running")
 			}
+		}
+	}
+}
+
+func (s *Scheduler) Start(id JobID) error {
+	job, ok := s.jobs[id]
+	if !ok {
+		return xerrors.Errorf("starting worker ID: %d not found", id)
+	}
 
-			// Keep this task running forever
-			doneFirstRun := false
-			for {
+	if job.running {
+		return xerrors.Errorf("starting worker ID: %d already running", id)
+	}
+	job.log.Info("starting job")
+	go s.execute(job, s.workerJobComplete)
+	return nil
+}
 
-				// Is the context done?
-				select {
-				case <-ctx.Done():
-					return
-				default:
-				}
+func (s *Scheduler) Stop(id JobID) error {
+	job, ok := s.jobs[id]
+	if !ok {
+		return xerrors.Errorf("starting worker ID: %d not found", id)
+	}
 
-				if doneFirstRun {
-					log.Infow("restarting task", "task", tc.Name, "delay", tc.RestartDelay)
-					if tc.RestartDelay > 0 {
-						time.Sleep(tc.RestartDelay)
-					}
-				} else {
-					log.Infow("running task", "task", tc.Name)
-					doneFirstRun = true
-				}
+	if !job.running {
```

Comment:
Needs a mutex - or put it in an `IsRunning()` method on job that guards access with mutex